### ISSUE TITLE
```DataArray.sel``` can silently pick up the nearest point, even if it is far away and the query is out of bounds

### ISSUE THREAD

**Issue by jerabaul29** [pydata/xarray#8335]
### What is your issue?

@paulina-t (who found a bug caused by the behavior we report here in a codebase, where it was badly messing things up).

See the example notebook at https://github.com/jerabaul29/public_bug_reports/blob/main/xarray/2023_10_18/interp.ipynb .

---

## Problem

It is always a bit risky to interpolate / find the nearest neighbor to a query or similar, as bad things can happen if querying a value for a point that is outside of the area that is represented. Fortunately, xarray returns NaN if performing ```interp``` outside of the bounds of a dataset:

```python
import xarray as xr
import numpy as np

xr.__version__

'2023.9.0'

data = np.array([[1, 2, 3], [4, 5, 6]])
lat = [10, 20]
lon = [120, 130, 140]

data_xr = xr.DataArray(data, coords={'lat':lat, 'lon':lon}, dims=['lat', 'lon'])

data_xr

<xarray.DataArray (lat: 2, lon: 3)>
array([[1, 2, 3],
       [4, 5, 6]])
Coordinates:
  * lat      (lat) int64 10 20
  * lon      (lon) int64 120 130 140

# interp is civilized: rather than wildly extrapolating, it returns NaN

data_xr.interp(lat=15, lon=125)

<xarray.DataArray ()>
array(3.)
Coordinates:
    lat      int64 15
    lon      int64 125

data_xr.interp(lat=5, lon=125)

<xarray.DataArray ()>
array(nan)
Coordinates:
    lat      int64 5
    lon      int64 125
```

Unfortunately, ```.sel``` will happily find the ```nearest``` neighbor of a point, even if the input point is outside of the dataset range:

```python
# sel is not as civilized: it happily finds the neares neighbor, even if it is "on the one side" of the example data

data_xr.sel(lat=5, lon=125, method='nearest')

<xarray.DataArray ()>
array(2)
Coordinates:
    lat      int64 10
    lon      int64 130
```

This can easily cause tricky bugs.

---

## Discussion

Would it be possible for ```.sel``` to have a behavior that makes the user aware of such issues? I.e. either:

- print a warning on stderr
- return NaN
- raise an exception

when performing a ```.sel``` query that is outside of a dataset range / not in between of 2 dataset points?

I understand that finding the nearest neighbor may still be useful / wanted in some cases even when being outside of the bounds of the dataset, but the fact that this happens silently by default has been causing bugs for us. Could either this default behavior be changed, or maybe enabled with a flag (```allow_extrapolate=False``` by default for example, so users can consciously opt it in)?

**Comment by max-sixty** [comment]
To confirm, this is limited to the case when `method="nearest"` is passed; is that correct?

How about adding a `method="interp"` to `.sel`? Then that's clearly discriminated from `method="nearest"`, which would continue to extrapolate.

**Comment by max-sixty** [comment]
Yes, sorry, my message was unclear. It could be named `method="nearest-surrounded"` (or some better name).

My point was that we could offer an option to explicitly select internal points, rather than change the meaning of the existing method, or raise warnings, since I'd guess that some folks do want to allow selecting edges.

**Comment by max-sixty** [comment]
I hadn't thought of this behavior as likely a bug. I also don't use this myself that much, so interested in what others think / whether you're confident here.

**Comment by max-sixty** [comment]
> For example looking up lat=-120 while lat range was 0...360 and not -180...180 in our case.

Yes, it makes sense that this is not intended.

But can we write down a rule that is general enough to handle all cases without much context? Since selecting the nearest value to 8 on `(9, 18, 27)` is probably not a bug.

**Comment by keewis** [comment]
I don't think we can do a lot here (correct me if I'm wrong, though, @benbovy): the `method` kwarg is handled by the [index implementation](https://github.com/pydata/xarray/blob/1e8f618e9d3754cb892a8438a1a7745aadce2bbf/xarray/core/indexes.py#L561) (in this case, a `pandas.Index`), so we'd need to convince `pandas` to implement this.

**Comment by kmuehlbauer** [comment]
I agree with @keewis. Xarray is just using plain pandas [Index.get_indexer](https://pandas.pydata.org/docs/reference/api/pandas.Index.get_indexer.html#pandas.Index.get_indexer) method under the hood.

@jerabaul29 To avoid pulling in data which is totally off (lat=-120, with lat-range 0..360) you can utilize `tolerance`-kwarg. You'd need to set it to grid_spacing/2 (5 in your case) to select any values using `method="nearest"`. This will still select points `outside` your data range, but only as far away as grid_spacing/2 and will avoid totally off errors. 

This might be solved by something like a `PeriodicBoundaryIndex`, where @benbovy is the right person to get involved.

Q: Might be a silly question, where are lat-ranges 0..360 or -180..-180 actually used?

**Comment by benbovy** [comment]
In theory we could implement some logic in the xarray `PandasIndex` wrapper such that it departs from `pandas.Index`, although I don't think it is always a good idea. In this case I agree with @kmuehlbauer  using `tolerance` might be good enough?

An alternative index maybe inheriting from `PandasIndex` may also be a solution that avoids cluttering the API / behavior of the default index (if you look at the implementation of `PandasIndex.sel` it is already quite complicated).

**Comment by jenseva** [comment]
I had a similar experience today when testing sel with longitude. I was able to see that sel returned an unexpected result and tracked it down to the mismatched longitude systems. It would have been nice to have a warning message. +1 for that offer.

**Comment by Mixpap** [comment]
You can use the tolerance keyword
```
data_xr.sel(lat=5, lon=125, method='nearest', tolerance=2) 
```
However, the problem with the tolerance is that it does not correspond to each coordinate, which means that you need to make sequential "sel" calls for each coordinate.
